template/runtime/src/lib.rs

```diff
@@ -441,9 +443,73 @@ impl_runtime_apis! {
 	impl sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block> for Runtime {
 		fn validate_transaction(
 			source: TransactionSource,
-			tx: <Block as BlockT>::Extrinsic,
+			xt: <Block as BlockT>::Extrinsic,
 		) -> TransactionValidity {
-			Executive::validate_transaction(source, tx)
+			// This template uses Substrate's pallet transaction payment. This
+			// makes the chain feel much more like a standard Substrate chain when submitting
+			// frame transactions and using Substrate ecosystem tools. It has the downside that
+			// transaction are not prioritized by gas_price. The following code reprioritizes
+			// transactions to overcome this.
+			//
+			// For chains that seek a more ethereum-first approach, a more elegant solution
+			// would be a pallet that replaces pallet transaction payment, and allows users
+			// to directly specify a gas price rather than computing an effective one.
+			// #HopefullySomeday
+
+			// First we pass the transactions to the standard FRAME executive. This calculates all the
+			// necessary tags, longevity and other properties that we will leave unchanged.
+			// This also assignes some priority that we don't care about and will overwrite next.
+			let mut intermediate_valid = Executive::validate_transaction(source, xt.clone())?;
+
+			// If this is a pallet ethereum transaction, then its priority is already set
+			// according to gas price from pallet ethereum. If it is any other kind of transaction,
+			// we modify its priority.
+			Ok(match &xt.function {
+				Call::Ethereum(transact(_)) => intermediate_valid,
+				_ => {
+					let dispatch_info = xt.get_dispatch_info();
+
+					// Try to extract the tip starting from https://crates.parity.io/sp_runtime/generic/struct.UncheckedExtrinsic.html
+					let tip = match xt.signature {
+						None => 0,
+						Some((_, _, signed_extra)) => {
+							// Yuck, this depends on the index of the charge transaction in this runtime's Signed Extra
+							let charge_transaction = signed_extra.6;
+							// Based on clues from:
+							// https://github.com/paritytech/substrate/blob/master/frame/transaction-payment/src/lib.rs#L477
+							// https://github.com/paritytech/substrate/blob/master/frame/transaction-payment/src/lib.rs#L501
+
+							// Dang, this didn't work
+							//   error[E0616]: field `0` of struct `ChargeTransactionPayment` is private
+							//      --> /home/joshy/ProgrammingProjects/frontier/template/runtime/src/lib.rs:481:37
+							//      |
+							//  481 | ...                   let tip = charge_transaction.0;
+							//      |                                                    ^ private field
+							let tip = charge_transaction.0;
+
+							tip
+						}
+					};
```

Comment:
@kianenigma Could help me out here?

We're trying to calculate an "effective gas price" for normal substrate transactions so they can be properly prioritized along with the ethereum ones.

The trouble I'm having is that I can't figure out how to get the amount of tip that the user has attached.

Reply:
seems like you need to either add a getter fn, or make the field pub, in downstream substrate

Reply:
https://github.com/paritytech/substrate/pull/9219